src/main/java/http/Router.java

```diff
@@ -1,23 +1,56 @@
 package http;
 
+import clientApplication.DynamicMethodNotAllowed;
+import clientApplication.DynamicOptions;
 import clientApplication.NotFound;
 
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public class Router implements Application {
     public Map<RouteId, Application> routes;
-    public Router(Map<RouteId, Application> routes){
+
+    public Router(Map<RouteId, Application> routes) {
         this.routes = routes;
     }
 
-    public HttpResponse start(HttpRequest clientRequest){
+    public HttpResponse start(HttpRequest clientRequest) {
         RouteId id = new RouteId(clientRequest.getMethod(), clientRequest.getRequestPath());
+        Application verifiedApplication = routes.get(id);
 
-        Optional<Application> requestedApplication = Optional.ofNullable(routes.get(id));
-        Application verifiedApplication = requestedApplication.orElse(new NotFound());
-        HttpResponse applicationResponse = verifiedApplication.start(clientRequest);
+        if (verifiedApplication == null) {
+            verifiedApplication = new NotFound();
+
+            if (pathExists(id.path)) {
+                String verbs = findVerbsForRequestPath(id);
+                verifiedApplication = new DynamicMethodNotAllowed(verbs);
 
+                if (id.verb == HttpMethod.OPTIONS) {
+                    verifiedApplication = new DynamicOptions(verbs);
+                }
+            }
+        }
```

Comment:
How can we make this more intention revealing? (This also violates one of our constraints I believe... 🤔)

Reply:
It does! 

However, I would argue since this bit of code creates an action that responds to a route type (like options, an undefined route or path) versus a particular route (method verb + path), it should be allowed to break the constraint. Also, `new NotFound()` was allowed in a previous story, so I figured it the other two actions would be fair play. 

In regards to intention, I could separate the conditionals into a function.

Reply:
>However, I would argue since this bit of code creates an action that responds to a route type (like options, an undefined route or path) versus a particular route (method verb + path), it should be allowed to break the constraint.

😄 Which constraint are you referring to? I was thinking of number of lines in a method. 🤓

Reply:
Ah, I was thinking of the `new` constraint. 

For that one, yeah, I should extract the if conditionals into a function lol.

Reply:
You could probably leverage the `.getOrDefault` method on HashMap, if you extract the logic for determining the default to another method.